### PR TITLE
Add jax.linear_transpose

### PR DESCRIPTION
**Pull request by shoyer**
This is intended as a relatively low-level building block to facilitate using functional transposes in user code.

Questions:

- are we happy with this name? Other options: `transpose_fun`,  `transpose` (but could be confused with `jax.numpy.transpose`)
- add a formal definition of what a "linear function" is? (Suggestions would be appreciated!)

Co-authored-by: Matthew Johnson <mattjj@google.com>

### CODE REVIEW

File: jax/api.py [hunk 1]
```diff
@@ -1575,6 +1574,56 @@ def _vjp(fun: lu.WrappedFun, *primals, **kwargs):
     return out_primal_py, vjp_py, tree_unflatten(aux_tree, aux)
 
 
+def linear_transpose(fun: Callable, *args) -> Callable:
+  """Transpose a function that is promised to be linear.
+
+  For linear functions, this transformation is equivalent to ``vjp``, but
+  avoids the overhead of overhead of computing the forward pass.
```
Comment (NeilGirdhar) [inline jax/api.py:1581]: FYI there's an error in the docstring :)

Comment (shoyer) [reply jax/api.py]: good catch, fixed!

**Comment by shoyer** [comment]
I would love to finish this up. Where I got stuck on is what exactly it means to transpose a "linear" function, particularly in the case of transposing a float -> complex or int -> float function.

I do care about transposing float -> complex functions (namely, rfft), but maybe I can raise an error for int -> float functions? Or maybe it's not worth worrying about too much, given how rare that sort of thing should be?

**Comment by apaszke** [comment]
There's recently [been some discussion](https://github.com/google/jax/issues/3402#issuecomment-644318023) relevant to the complex transpostion.

**Comment by ibab** [comment]
Hi @shoyer, I've found your function to be super useful!
I'm actually dealing with a complex->complex function and the transformation is giving correct answers for me.
I think throwing an error in the case of `int` would make sense.

Have you thought about changing the API of the function so that the arguments can be passed in at a later time? E.g.
```python
transpose_fn = jax.linear_transpose(my_func)
transpose_fn(*args)
```
This can be done at the cost of re-tracing the function each time it's called. Since I `jit` the function anyway this was more convenient for me.

One issue that I've encountered is that the transformed function will return a tuple, even in the case where the function takes and returns a single tensor. I kind of understand where that comes from (the input spec becomes the output spec), but I think it would be nice to treat the case of functions with a single input in in a special way (unpack the output of the transpose).

Also it's very useful to define the Hermitian:
```python
def linear_hermitian(fn):
  def transformed_fn(*args):
    args = jax.tree_map(lambda x: jnp.conj(x), args)
    outs = linear_transpose(fn, *args)(*args)
    return jax.tree_map(lambda x: jnp.conj(x), outs)
  return transformed_fn
```

**Comment by shoyer** [comment]
OK, I think this is ready for review. For now I've taken @ibab's suggestion of simply not support integer dtypes, for a few reasons:

1. It seems unlikely that the transpose of integer -> float functions is being calculated correctly.
2. We don't have any use-cases for even integer -> integer functions yet. It is always easier to turn errors into supported features rather than the other way around.
3. There is potentially some conflict with defining the gradient of integer valued functions as zero (https://github.com/google/jax/pull/4039).

**Review by mattjj — APPROVED** [review]
Love it, thank you so much! You put in a lot of work after our initial pair programming session :)
